src/core/CsvParser.cpp

```diff
@@ -0,0 +1,411 @@
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <QTextCodec>
+#include <QObject>
+#include "core/Tools.h"
+#include "CsvParser.h"
+
+CsvParser::CsvParser()
+    : m_ch(0)
+    , m_comment('#')
+    , m_currCol(1)
+    , m_currRow(1)
+    , m_isBackslashSyntax(false)
+    , m_isEof(false)
+    , m_isFileLoaded(false)
+    , m_isGood(true)
+    , m_lastPos(-1)
+    , m_maxCols(0)
+    , m_qualifier('"')
+    , m_separator(',')
+    , m_statusMsg("")
+{
+    m_csv.setBuffer(&m_array);
+    m_ts.setDevice(&m_csv);
+    m_csv.open(QIODevice::ReadOnly);
+    m_ts.setCodec("UTF-8");
+}
+
+CsvParser::~CsvParser() {
+    m_csv.close();
+}
+
+bool CsvParser::isFileLoaded() {
+    return m_isFileLoaded;
+}
+
+bool CsvParser::reparse() {
+    reset();
+    return parseFile();
+}
+
+
+bool CsvParser::parse(QFile *device) {
+    clear();
+    if (nullptr == device) {
+        m_statusMsg += QObject::tr("NULL device\n");
+        return false;
+    }
+    if (!readFile(device)) {
+        return false;
+    }
+    return parseFile();
+}
+
+bool CsvParser::readFile(QFile *device) {
+    if (device->isOpen()) {
+        device->close();
+    }
+
+    device->open(QIODevice::ReadOnly);
+    if (!Tools::readAllFromDevice(device, m_array)) {
+        m_statusMsg += QObject::tr("Error reading from device\n");
+        m_isFileLoaded = false;
+    }
+    else {
+    device->close();
+
+    m_array.replace("\r\n", "\n");
+    m_array.replace("\r", "\n");
+    if (0 == m_array.size()) {
+        m_statusMsg += QObject::tr("File empty\n");
+    }
+    m_isFileLoaded = true;
+    }
+    return m_isFileLoaded;
+}
+
+void CsvParser::reset() {
+    m_ch = 0;
+    m_currCol = 1;
+    m_currRow = 1;
+    m_isEof = false;
+    m_isGood = true;
+    m_lastPos = -1;
+    m_maxCols = 0;
+    m_statusMsg = "";
+    m_ts.seek(0);
+    m_table.clear();
+    //the following are users' concern :)
+    //m_comment = '#';
+    //m_backslashSyntax = false;
+    //m_comment = '#';
+    //m_qualifier = '"';
+    //m_separator = ',';
+}
+
+void CsvParser::clear() {
+    reset();
+    m_isFileLoaded = false;
+    m_array.clear();
+}
+
+bool CsvParser::parseFile() {
+    parseRecord();
+    while (!m_isEof)
+    {
+        if (!skipEndline()) {
+            appendStatusMsg(QObject::tr("malformed string"));
+        }
+        m_currRow++;
+        m_currCol = 1;
+        parseRecord();
+    }
+    fillColumns();
+    return m_isGood;
+}
+
+void CsvParser::parseRecord() {
+    csvrow row;
+    if (isComment()) {
+        skipLine();
+        return;
+    }
+    else {
```

Comment:
Return is expected only when `if (isComment())` is true. I think you made a typo :eyes: